[DOCS] Data stream modification API (#80094)
elastic/elasticsearch: docs/reference/data-streams/modify-data-streams-api.asciidoc

[[modify-data-streams-api]]
=== Modify data streams API
++++
<titleabbrev>Modify data streams API</titleabbrev>
++++

Performs one or more data stream modification actions in a single atomic
operation.

////
[source,console]
----
PUT /_index_template/template
{
  "index_patterns": ["my-logs*"],
  "data_stream": { }
}

POST my-logs/_doc/
{
    "message" : "trying out DS modification",
    "@timestamp": "2021-01-01"
}

POST my-logs/_rollover

POST index-to-add/_doc/
{
    "description" : "this is the index to be added to a data stream",
    "@timestamp": "2021-01-01"
}
----
////

[source,console]
----
POST _data_stream/_modify
{
  "actions": [
    {
      "remove_backing_index": {
        "data_stream": "my-logs",

>         "index": ".ds-my-logs-2021.10.15-000001"

Author: @debadair, the specific name of the data stream's backing index here is causing one of the doc integration tests to fail. I believe we have other docs such as the "Use a data stream" page where the names of date-dependent backing indices are used, but I don't know how those were set up to be tested successfully.

Reviewer: The example on the use a data stream page sets up a not-timestamped index and substitutes that index name for the timestamped one. That's not going to work for this call, though. If you swap the actions, you could remove the index you add and the test should work. I'd call it something other than "index-to-add" if you do that. The other option would be to skip the test.

      }
    },
    {
      "add_backing_index": {
        "data_stream": "my-logs",
        "index": "index-to-add"
      }
    }
  ]
}
----
// TEST[continued]

////
[source,console]
-----------------------------------
DELETE /_data_stream/my-logs
DELETE /_index_template/template
-----------------------------------
// TEST[continued]
////

[[modify-data-streams-api-request]]
==== {api-request-title}

`POST /_data_stream/_modify`

[role="child_attributes"]
[[modify-data-streams-api-request-body]]
==== {api-request-body-title}

`actions`::
(Required, array of objects) Actions to perform.
+
.Properties of `actions` objects
[%collapsible%open]
====
`<action>`::
(Required, object) The key is the action type. At least one action is required.
+
.Valid `<action>` keys
[%collapsible%open]
=====
`add_backing_index`::

> Adds an existing index as backing index to a data stream. The index will be
> hidden as part of the operation.

Reviewer: ```suggestion
Adds an existing index as a backing index for a data stream. The index is
hidden as part of this operation.
```

> WARNING: Addition of certain indices with the `add_backing_index` operation
> can result in improper data stream behavior. This is a potentially
> destructive action and should be considered an expert level API.

Reviewer: Can you be more specific about what indices cause problems? Otherwise, I'd just drop the qualification. (It sounds scarier, but that's probably not a bad thing.)

```suggestion
WARNING: Adding indices with the `add_backing_index` action
can potentially result in improper data stream behavior. 
This should be considered an expert level API.
```

`remove_backing_index`::

> Removes a backing index from a data stream. The index will be unhidden
> as part of the operation. A data stream's write index cannot be removed.

Reviewer: ```suggestion
Removes a backing index from a data stream. The index is unhidden
as part of this operation. A data stream's write index cannot be removed.
```
